File: src/atomsciflow/post/post.cpp
/************************************************************************
MIT License

Copyright (c) 2022 Deqi Tang

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
************************************************************************/

#include "atomsciflow/post/post.h"

#include <iostream>
#include <chrono>
#include <thread>
#ifdef _OPENMP
#include <omp.h>
#endif // _OPENMP

namespace atomsciflow::post {

Post::Post() {
    this->set_run("post-dir", "post.dir");
    this->set_run("program-out", "program.out");
    this->set_run("output-json", "post-program.json");
}

Post::~Post() {

}

void Post::read_lines(const std::string& filepath) {
    std::ifstream stream;
    stream.open(filepath);
    std::string line;

    while (std::getline(stream, line)) {
        this->lines.emplace_back(line);
    }
    stream.close();
}

void Post::read(const std::string& filepath) {
    auto read_lines_start = std::chrono::system_clock::now();
    this->read_lines(filepath);
    std::chrono::duration<double> time_consumed = std::chrono::system_clock::now() - read_lines_start;
    std::cout << "Time for reading lines: " << time_consumed.count() << " seconds\n";

    auto apply_rules = [&](const std::string& line) {
        // #pragma omp parallel for
        // for (auto& item : this->rules) {
        //     boost::any_cast<std::function<void(const std::string&)>>(item)(line);
        // }
        auto apply_rules_partition = [&](int start, int end) {
            for (auto it = rules.begin() + start; it != rules.begin() + end; it++) {
                boost::any_cast<std::function<void(const std::string&)>>(*it)(line);
            }
        };
        int n_threads = 4;
        int n_rules = this->rules.size();
        int n_rules_each_task_base = n_rules / n_threads;
        int n_rules_remainder = n_rules % n_threads;
        std::vector<int> n_rules_each_task;
        for (int i = 0; i < n_threads; i++) {
            if (i < n_rules_remainder) {
                n_rules_each_task.push_back(n_rules_each_task_base + 1);
            } else {
                n_rules_each_task.push_back(n_rules_each_task_base);
            }
        }
        std::vector<std::thread> tasks;
        int rules_range = 0;
        for (int i = 0; i < n_threads; i++) {
            tasks.push_back(std::thread{
                apply_rules_partition,
                rules_range,
                rules_range + n_rules_each_task[i]
            });
            rules_range += n_rules_each_task[i];
        }
        for (auto& item : tasks) {
            item.join();
        }
    };

    auto apply_rules_start = std::chrono::system_clock::now();
    for (const auto& item : lines) {
        apply_rules(item);
    }
    time_consumed = std::chrono::system_clock::now() - apply_rules_start;
    std::cout << "Time to apply rules: " << time_consumed.count() << " seconds\n";
}

void Post::write(const std::string& directory) {
    pt::write_json((fs::path(directory) / run_params["output-json"]).string(), this->info);
}

void Post::set_run(std::string key, std::string value) {
    this->run_params[key] = value;
}

void Post::run(const std::string& directory) {
    this->read((fs::path(directory) / this->run_params["program-out"]).string());
    fs::create_directory(fs::path(directory) / run_params["post-dir"]);
    this->write((fs::path(directory) / run_params["post-dir"]).string());
}

void Post::add_rule(boost::any rule) {
    this->rules.emplace_back(rule);
}

void Post::add_rule_type_1(std::string pat1, std::string pat2) {
    this->add_rule(std::function<void(const std::string&)>{[&](const std::string& str) {
        std::regex re_pat1{pat1};
        std::regex re_pat2{pat2};
        std::smatch m1;
        std::smatch m2;
        if (std::regex_search(str, m1, re_pat1)) {
            std::regex_search(str, m2, re_pat2);
            this->info.put(m1.str(0), m2.str(0));
        }
    }});
}

} // namespace atomsciflow::post

File: atomsciflow/elk/post.py
"""
MIT License

Copyright (c) 2022 Deqi Tang

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""

import os

from atomsciflow.cpp import elk
from atomsciflow.cpp import post

class Opt(elk.PostOpt):
    def __init__(self):
        super().__init__()

class Band(post.Post):
    def __init__(self):
        super().__init__()

    def set_kpath(self, kpath):
        self.kpath = kpath

    def run(self, directory):
        super().run(directory)

        ha_to_ev = 27.211324570273

        with open(os.path.join(directory, "BANDLINES.OUT") ,"r") as fin:
            bandlines_out = fin.readlines()

        xtics_locs = []
        xtics_locs.append(float(bandlines_out[0].split()[0]))
        for i in range(1, len(bandlines_out)):
            if len(bandlines_out[i].split()) == 0:
                continue
            if (float(bandlines_out[i].split()[0]) - xtics_locs[-1]) < 1.0e-5:
                continue
            else:
                xtics_locs.append(float(bandlines_out[i].split()[0]))

        xtics_labels = []
        for label in self.kpath.labels:
            if label != "GAMMA":
                xtics_labels.append(label)
            else:
                xtics_labels.append("{/symbol G}")

        with open(os.path.join(directory, "post.dir/band.gnuplot"), "w") as fout:
            fout.write("set terminal png\n")
            fout.write("unset key\n")
            fout.write("set parametric\n")
            fout.write("set title 'Band structure' font ',15'\n")
            fout.write("set ylabel 'Energy (eV)' font ',15'\n")
            fout.write("set ytics font ',15'\n")
            fout.write("set xtics font ',15'\n")
            fout.write("set border linewidth 3\n")
            fout.write("set autoscale\n")
            # set a linestyle 1 to be the style for band lines
            fout.write("set style line 1 linecolor rgb \'black\' linetype 1 pointtype 1 linewidth 3\n")
            # set a linestyle 2 to be the style for the vertical high-symmetry kpoint line and horizontal fermi level line
            fout.write("set style line 2 linecolor rgb \'black\' linetype 1 pointtype 2 linewidth 0.5\n")
            
            fout.write("set xtics(")
            for i in range(len(xtics_locs) - 1):
                fout.write("\'%s\' %f, "% (xtics_labels[i], xtics_locs[i]))
            fout.write("\'%s\' %f)\n" % (xtics_labels[-1], xtics_locs[-1]))
            
            for x in xtics_locs:
                fout.write("set arrow from %f, graph 0 to %f, graph 1 nohead linestyle 2\n" % (x, x))
            fout.write("set arrow from 0, 0 to %f, 0 nohead linestyle 2\n" % xtics_locs[-1])

            fout.write("set output 'band-structure.png'\n")
            fout.write("plot \'../BAND.OUT\' using 1:(column(2)*%f) w l notitle linestyle 1\n" % (ha_to_ev))

        with open(os.path.join(directory, "post.dir/analysis.sh"), "w") as fout:
            fout.write("#!/bin/bash\n\n")
            fout.write("#\n")
            fout.write("cd %s\n" % os.path.abspath(os.path.join(directory, self.run_params["post-dir"])))
            fout.write("\n")
            fout.write("gnuplot band.gnuplot\n")

        cmd = ""
        cmd += "bash "
        cmd += os.path.join(directory, self.run_params["post-dir"] + "/analysis.sh")
        os.system(cmd)

class Dos(post.Post):
    def __init__(self):
        super().__init__()

    def run(self, directory):
        super().run(directory)
        import os
        from atomsciflow.cpp.base import Atom, Xyz

        ha_to_ev = 27.211324570273

        with open(os.path.join(directory, "elk.in"), "r") as fin:
            elk_in_lines = fin.readlines()
        atoms_elk = []
        for i in range(len(elk_in_lines)):
            if elk_in_lines[i].replace(" ", "").replace("\n", "") == "atoms":
                atoms_begin_index = i
                break
        nspecies = int(elk_in_lines[atoms_begin_index+1].split()[0])
        specie_i_start = atoms_begin_index + 2
        for i in range(nspecies):
            name = elk_in_lines[specie_i_start].split(".")[0].split("'")[1]
            natoms = int(elk_in_lines[specie_i_start+1].split()[0])
            elem = {}
            elem["element"] = name
            elem["atoms"] = []
            for j in range(natoms):
                atom = Atom()
                atom.name = name
                atom.x = float(elk_in_lines[specie_i_start+2+j].split()[0])
                atom.y = float(elk_in_lines[specie_i_start+2+j].split()[1])
                atom.z = float(elk_in_lines[specie_i_start+2+j].split()[2])
                elem["atoms"].append(atom)
            atoms_elk.append(elem)
            specie_i_start = specie_i_start + 1 + natoms + 1
        with open(os.path.join(directory, "post.dir/dos.gnuplot"), "w") as fout:
            fout.write("set terminal png\n")
            fout.write("set parametric\n")
            fout.write("set title 'Density of state' font ',15'\n")
            fout.write("set ylabel 'Density of state' font ',15'\n")
            fout.write("set xlabel 'Energy (eV)' font ',15'\n")
            fout.write("set ytics font ',15'\n")
            fout.write("set xtics font ',15'\n")
            fout.write("set border linewidth 3\n")
            fout.write("set autoscale\n")
            fout.write("set xrange [-10:10]\n")

            fout.write("set arrow from 0, graph 0 to 0, graph 1 nohead linecolor rgb \'black\' linewidth 0.5\n")
            fout.write("set arrow from -10, 0 to 10, 0 nohead linecolor rgb \'black\' linewidth 0.5\n")

            fout.write("set output 'total-dos.png'\n")
            fout.write("plot \'../TDOS.OUT\' using (column(1) * %f):2 w l notitle linewidth 3\n" % (ha_to_ev))
                                                 
            fout.write("set output 'atom-projected-dos.png\n")
            fout.write("plot \\\n")
            for i in range(len(atoms_elk)):
                for j in range(len(atoms_elk[i]["atoms"])):
                    fout.write("  '../PDOS_S%02d_A%04d.OUT' using (column(1) * %f):2 w l title '%s' linewidth 3,\\\n" % (
                        i+1,
                        j+1,
                        ha_to_ev,
                        "%s(%d)" % (atoms_elk[i]["element"], j+1)
                    ))
        with open(os.path.join(directory, "post.dir/analysis.sh"), "w") as fout:
            fout.write("#!/bin/bash\n\n")
            fout.write("#\n")
            fout.write("cd %s\n" % os.path.abspath(os.path.join(directory, self.run_params["post-dir"])))
            fout.write("\n")
            fout.write("gnuplot dos.gnuplot\n")

        cmd = ""
        cmd += "bash "
        cmd += os.path.join(directory, self.run_params["post-dir"] + "/analysis.sh")
        os.system(cmd)

class Phonopy(elk.PostPhonopy):
    def __init__(self):
        super().__init__()

    def run(self, directory):
        super().run(directory)
        self.extract_data(directory)
